Source/WebCore/accessibility/AXLiveRegionManager.cpp

```diff
@@ -0,0 +1,317 @@
+        return false;
+
+    // For leaf objects, include if they have a value (e.g., form controls).
+    if (!object.stringValue().isEmpty())
+        return true;
+
+#if PLATFORM(COCOA)
+    // For leaf objects, include if they have accessible description text (e.g., images with alt text).
+    if (!object.descriptionAttributeValue().isEmpty())
+        return true;
+#endif
+
+    return false;
+}
+
+String AXLiveRegionManager::textForObject(AccessibilityObject& object) const
+{
+    return object.textMarkerRange().toString(IncludeListMarkerText::Yes, IncludeImageAltText::Yes);
+}
+
+AXLiveRegionManager::LiveRegionDiff AXLiveRegionManager::computeChanges(const Vector<LiveRegionObject>& oldObjects, const Vector<LiveRegionObject>& newObjects) const
+{
+    // Here we compare the old and new live region to compute:
+    // - Additions: New objects
+    // - Deletions: Objects that were removed from the region
+    // - Changes: Text content/values that changed between the same object.
+
+    LiveRegionDiff diff;
+
+    // Build a map of old objects for lookup. As we match them with new objects, we'll remove them.
+    // Whatever remains unmatched at the end represents removals.
+    HashMap<AXID, String> unmatchedOldObjects;
+    for (auto& object : oldObjects)
+        unmatchedOldObjects.set(object.objectID, object.text);
+
+    for (auto& newObject : newObjects) {
+        auto it = unmatchedOldObjects.find(newObject.objectID);
+        if (it == unmatchedOldObjects.end())
+            diff.added.append(newObject);
+        else {
+            if (it->value != newObject.text)
+                diff.changed.append(newObject);
+            unmatchedOldObjects.remove(it);
+        }
+    }
+
+    for (auto& entry : unmatchedOldObjects)
+        diff.removed.append({ entry.key, entry.value });
+
+    return diff;
+}
+
+String AXLiveRegionManager::computeAnnouncement(const LiveRegionSnapshot& newSnapshot, const LiveRegionDiff& diff) const
+{
+    bool hasAll = newSnapshot.liveRegionRelevant.contains(LiveRegionRelevant::All);
+    bool hasAdditions = hasAll || newSnapshot.liveRegionRelevant.contains(LiveRegionRelevant::Additions);
+    bool hasRemovals = hasAll || newSnapshot.liveRegionRelevant.contains(LiveRegionRelevant::Removals);
+    bool hasText = hasAll || newSnapshot.liveRegionRelevant.contains(LiveRegionRelevant::Text);
+
+    Vector<String> strings;
+
+    if (hasAdditions && !diff.added.isEmpty()) {
+        for (auto& object : diff.added) {
+            if (!object.text.isEmpty())
+                strings.append(object.text);
+        }
+    }
+
+    if (hasRemovals && !diff.removed.isEmpty()) {
+        StringBuilder removalString;
+        removalString.append(AXRemovedText());
+        for (auto& object : diff.removed) {
+            if (!object.text.isEmpty()) {
+                removalString.append(' ');
+                removalString.append(object.text);
+            }
+        }
+        strings.append(removalString.toString());
```

Comment:
It looks like it's possible for diff.removed to be nonempty, but all of the text is empty, so you may not want to append to strings in that case